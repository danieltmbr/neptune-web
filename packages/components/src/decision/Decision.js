import React from 'react';
import Types from 'prop-types';
import SizeSwapper from '../sizeswapper';
import NavigationOption from '../navigationOption';
import Tile from '../tile';
import { elementOfType } from '../common/customProps';

const DecisionType = { NAVIGATION: 'NAVIGATION' };
const DecisionPresentantion = {
  LIST: 'LIST',
  LIST_BLOCK: 'LIST_BLOCK',
  LIST_BLOCK_SMALL: 'LIST_BLOCK_SMALL',
};

const Decision = ({ presentation, options, type }) => {
  if (type === DecisionType.NAVIGATION) {
    const { LIST_BLOCK, LIST_BLOCK_SMALL } = DecisionPresentantion;
    if (presentation === LIST_BLOCK || LIST_BLOCK_SMALL) {
      const small = presentation === LIST_BLOCK_SMALL;
      return options.map(({ title, description, media: { list, block }, onClick }, key) => (
        <SizeSwapper
          // eslint-disable-next-line react/no-array-index-key
          key={key}
          items={[
            <NavigationOption
              complex={false}
              href="#"
              title={title}
              description={description}
              onClick={onClick}
              media={list}
              showMediaAtAllSizes
              showMediaCircle
            />,
            <Tile
              title={title}
              description={description}
              media={block}
              size={small ? Tile.Size.SMALL : null}
              onClick={onClick}
            />,
          ]}
        />
      ));
    }
    // LIST
    return options.map(({ title, description, media: { list }, onClick }, key) => (
      <NavigationOption
        // eslint-disable-next-line react/no-array-index-key
        key={key}
        complex={false}
        href="#"
        title={title}
        description={description}
        onClick={onClick}
        media={list}
        showMediaAtAllSizes
        showMediaCircle
      />
    ));
  }
  return <></>;
};

Decision.Presentation = DecisionPresentantion;
Decision.Type = DecisionType;

Decision.propTypes = {
  /**  Handles the display mode of the component. */
  presentation: Types.oneOf([
    Decision.Presentation.LIST,
    Decision.Presentation.LIST_BLOCK,
    Decision.Presentation.LIST_BLOCK_SMALL,
  ]),
  /**  The options to be rendered. */
  options: Types.arrayOf(
    Types.shape({
      media: Types.shape({
        list: elementOfType(['Avatar', 'img']),
        block: elementOfType(['Avatar', 'img']),
      }),
      title: Types.string.isRequired,
      description: Types.node.isRequired,
      onClick: Types.func.isRequired,
    }),
  ).isRequired,
  /** The element type to be rendered. */
  type: Types.oneOf([Decision.Type.NAVIGATION]),
};

Decision.defaultProps = {
  presentation: Decision.Presentation.LIST,
  type: Decision.Type.NAVIGATION,
};

export default Decision;
