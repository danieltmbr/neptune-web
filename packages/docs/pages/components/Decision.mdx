import { LiveEditorBlock, GeneratePropsTable } from '../../utils';
import { Decision } from '@transferwise/components';
import code from '../../liveEditorCode/Decision.code';

<LiveEditorBlock code={code} scope={{ Decision }} />
<GeneratePropsTable componentName="Decision" />

export const meta = {
  name: 'Decision',
};