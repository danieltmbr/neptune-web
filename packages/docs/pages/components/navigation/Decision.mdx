import { LiveEditorBlock, GeneratePropsTable } from '../../../utils';
import code from '../../../liveEditorCode/decision.code.js';

import { Decision, Avatar, Tile, NavigationOption } from '@transferwise/components';

<LiveEditorBlock code={code} scope={{ Decision, Avatar }} display="vertical" />
<GeneratePropsTable componentName="Decision" />

## Usage

Decision can be used for illustrating major decisions for the user.
It is encouraged to keep the number of options in `items` under three.

The `type` prop handles the kind of elements to display for each presentation style

### Navigation type

<table className="docs-table table table-condensed">
  <thead>
    <tr>
      <th>Presentation</th>
      <th>Mobile</th>
      <th>Desktop</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>{Decision.Presentation.LIST}</td>
      <td>
        <NavigationOption
          id="navigation-option"
          name="navigation-option"
          className="custom-navigation-option"
          title="Navigation option"
          content="Button and icon are vertically centered."
          onClick={() => alert('clicked')}
          media={
            <Avatar size="md" theme="light" type="initials" className="text-xs-center">
              HM
            </Avatar>
          }
          showMediaAtAllSizes={false}
          showMediaCircle={true}
        />
      </td>
      <td>
        <NavigationOption
          id="navigation-option"
          name="navigation-option"
          className="custom-navigation-option"
          title="Navigation option"
          content="Button and icon are vertically centered."
          onClick={() => alert('clicked')}
          media={
            <Avatar size="md" theme="light" type="initials" className="text-xs-center">
              HM
            </Avatar>
          }
          showMediaAtAllSizes={false}
          showMediaCircle={true}
        />
      </td>
    </tr>
    <tr>
      <td>{Decision.Presentation.LIST_BLOCK}</td>
      <td>
        <NavigationOption
          id="navigation-option"
          name="navigation-option"
          className="custom-navigation-option"
          title="Navigation option"
          content="Button and icon are vertically centered."
          onClick={() => alert('clicked')}
          media={
            <Avatar size="md" theme="light" type="initials" className="text-xs-center">
              HM
            </Avatar>
          }
          showMediaAtAllSizes={false}
          showMediaCircle={true}
        />
      </td>
      <td>
        <Tile
          media={
            <img
              src="https://transferwise.com/public-resources/assets/bank-details/bank-details-flow/finish.svg"
              alt="illustration"
              style={{ width: '100%' }}
            />
          }
          size={Tile.Size.LARGE}
          title="Send money"
          content="Content"
          onClick={() => alert('onClick')}
        />
      </td>
    </tr>
    <tr>
      <td>{Decision.Presentation.LIST_BLOCK_SMALL}</td>
      <td>
        <NavigationOption
          id="navigation-option"
          name="navigation-option"
          className="custom-navigation-option"
          title="Navigation option"
          content="Button and icon are vertically centered."
          onClick={() => alert('clicked')}
          media={
            <Avatar size="md" theme="light" type="initials" className="text-xs-center">
              HM
            </Avatar>
          }
          showMediaAtAllSizes={false}
          showMediaCircle={true}
        />
      </td>
      <td>
        <Tile
          media={
            <img
              src="https://transferwise.com/public-resources/assets/bank-details/bank-details-flow/finish.svg"
              alt="illustration"
              style={{ width: '100%' }}
            />
          }
          size={Tile.Size.SMALL}
          title="Send money"
          content="Content"
          onClick={() => alert('onClick')}
        />
      </td>
    </tr>
  </tbody>
</table>

export const meta = {
  name: 'Decision',
};
