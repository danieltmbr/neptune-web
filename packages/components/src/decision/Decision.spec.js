import React from 'react';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import Decision from '.';
import Avatar from '../avatar';

const originalClientWidth = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'clientWidth');
const resetClientWidth = (width) => {
  Object.defineProperty(HTMLElement.prototype, 'clientWidth', { configurable: true, value: width });
};

const props = {
  type: Decision.Type.NAVIGATION,
  presentation: Decision.Presentation.LIST_BLOCK,
  options: [
    {
      media: {
        list: <Avatar type="initials">HM</Avatar>,
        block: <img src="img.jpg" alt="alt" />,
      },
      title: 'title',
      description: 'description',
      onClick: jest.fn(),
    },
  ],
};

describe('Decision', () => {
  afterAll(() => {
    Object.defineProperty(HTMLElement.prototype, 'clientWidth', originalClientWidth);
  });

  describe(`when presentation is ${Decision.Presentation.LIST_BLOCK}`, () => {
    let container;
    beforeEach(() => {
      ({ container } = render(<Decision {...props} />));
    });

    it('renders Navigation Option under first breakpoint', () => {
      expect(container).toMatchSnapshot();
    });
    it('renders Tile passed first breakpoint', () => {
      resetClientWidth(900);
      expect(container).toMatchSnapshot();
    });
  });

  describe(`when presentation is ${Decision.Presentation.LIST_BLOCK_SMALL}`, () => {
    let container;
    beforeEach(() => {
      ({ container } = render(
        <Decision {...props} presentation={Decision.Presentation.LIST_BLOCK_SMALL} />,
      ));
    });

    it('renders Navigation Option under first breakpoint', () => {
      expect(container).toMatchSnapshot();
    });
    it('renders Small Tile passed first breakpoint', () => {
      resetClientWidth(900);
      expect(container).toMatchSnapshot();
    });
  });

  describe(`when presentation is ${Decision.Presentation.LIST}`, () => {
    let container;
    beforeEach(() => {
      ({ container } = render(<Decision {...props} presentation={Decision.Presentation.LIST} />));
    });

    it('renders Navigation Option under first breakpoint', () => {
      expect(container).toMatchSnapshot();
    });
    it('renders Navigation Option passed first breakpoint', () => {
      resetClientWidth(900);
      expect(container).toMatchSnapshot();
    });
  });
});
