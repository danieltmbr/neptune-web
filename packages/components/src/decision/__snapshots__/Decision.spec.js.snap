// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decision when presentation is LIST renders Navigation Option passed first breakpoint 1`] = `
<div>
  <div
    class="tw-size-swapper"
  >
    <div
      class="tw-tile d-flex flex-column text-xs-center tw-tile--large p-y-5 p-x-4"
    >
      <div
        class="tw-tile__media m-x-auto"
      >
        <img
          alt="alt"
          src="img.jpg"
        />
      </div>
      <a
        class="tw-tile__link text-no-decoration "
        role="button"
        tabindex="0"
      >
        <span
          class="h4"
        >
          title
        </span>
      </a>
      <div
        class="tw-tile__content"
      >
        content
      </div>
    </div>
  </div>
</div>
`;

exports[`Decision when presentation is LIST renders Navigation Option under first breakpoint 1`] = `
<div>
  <div
    class="tw-size-swapper"
  >
    <div
      class="tw-tile d-flex flex-column text-xs-center tw-tile--large p-y-5 p-x-4"
    >
      <div
        class="tw-tile__media m-x-auto"
      >
        <img
          alt="alt"
          src="img.jpg"
        />
      </div>
      <a
        class="tw-tile__link text-no-decoration "
        role="button"
        tabindex="0"
      >
        <span
          class="h4"
        >
          title
        </span>
      </a>
      <div
        class="tw-tile__content"
      >
        content
      </div>
    </div>
  </div>
</div>
`;

exports[`Decision when presentation is LIST_BLOCK renders Navigation Option under first breakpoint 1`] = `
<div>
  <div
    class="tw-size-swapper"
  >
    <a
      class="tw-navigation-option media decision tw-option__sm-media"
      href="#"
    >
      <div
        class="media-left"
      >
        <div
          class="circle circle-sm text-primary circle-inverse"
        >
          <div
            class="tw-avatar tw-avatar--md tw-avatar--initials tw-avatar--light"
          >
            <div
              class="tw-avatar__content"
            >
              HM
            </div>
          </div>
        </div>
      </div>
      <div
        class="media-body"
      >
        <div
          class="h5"
        >
          title
        </div>
        content
      </div>
      <div
        class="media-right"
      >
        <span
          aria-hidden="true"
          class="tw-icon tw-icon-chevron-up tw-chevron chevron-color right"
          role="presentation"
        >
          <svg
            fill="currentColor"
            height="16"
            width="16"
          >
            <path
              clip-rule="evenodd"
              d="M8 4l-6.6 6.653L2.537 11.8 8 6.293l5.463 5.507 1.137-1.147L8 4z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </a>
  </div>
</div>
`;

exports[`Decision when presentation is LIST_BLOCK renders Tile passed first breakpoint 1`] = `
<div>
  <div
    class="tw-size-swapper"
  >
    <a
      class="tw-navigation-option media decision tw-option__sm-media"
      href="#"
    >
      <div
        class="media-left"
      >
        <div
          class="circle circle-sm text-primary circle-inverse"
        >
          <div
            class="tw-avatar tw-avatar--md tw-avatar--initials tw-avatar--light"
          >
            <div
              class="tw-avatar__content"
            >
              HM
            </div>
          </div>
        </div>
      </div>
      <div
        class="media-body"
      >
        <div
          class="h5"
        >
          title
        </div>
        content
      </div>
      <div
        class="media-right"
      >
        <span
          aria-hidden="true"
          class="tw-icon tw-icon-chevron-up tw-chevron chevron-color right"
          role="presentation"
        >
          <svg
            fill="currentColor"
            height="16"
            width="16"
          >
            <path
              clip-rule="evenodd"
              d="M8 4l-6.6 6.653L2.537 11.8 8 6.293l5.463 5.507 1.137-1.147L8 4z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </a>
  </div>
</div>
`;

exports[`Decision when presentation is LIST_BLOCK_SMALL renders Navigation Option under first breakpoint 1`] = `
<div>
  <div
    class="tw-size-swapper"
  >
    <div
      class="tw-tile d-flex flex-column text-xs-center tw-tile--small p-a-2"
    >
      <div
        class="tw-tile__media m-x-auto"
      >
        <img
          alt="alt"
          src="img.jpg"
        />
      </div>
      <a
        class="tw-tile__link text-no-decoration "
        role="button"
        tabindex="0"
      >
        <span
          class="h5"
        >
          title
        </span>
      </a>
      <div
        class="tw-tile__content"
      >
        content
      </div>
    </div>
  </div>
</div>
`;

exports[`Decision when presentation is LIST_BLOCK_SMALL renders Small Tile passed first breakpoint 1`] = `
<div>
  <div
    class="tw-size-swapper"
  >
    <div
      class="tw-tile d-flex flex-column text-xs-center tw-tile--small p-a-2"
    >
      <div
        class="tw-tile__media m-x-auto"
      >
        <img
          alt="alt"
          src="img.jpg"
        />
      </div>
      <a
        class="tw-tile__link text-no-decoration "
        role="button"
        tabindex="0"
      >
        <span
          class="h5"
        >
          title
        </span>
      </a>
      <div
        class="tw-tile__content"
      >
        content
      </div>
    </div>
  </div>
</div>
`;
